src/nomad_simulations/properties/band_structure.py

```diff
@@ -0,0 +1,377 @@
+        """,
+    )
+
+    highest_occupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Highest occupied electronic eigenvalue. Together with `lowest_unoccupied`, it defines the
+        electronic band gap.
+        """,
+    )
+
+    lowest_unoccupied = Quantity(
+        type=np.float64,
+        unit='joule',
+        description="""
+        Lowest unoccupied electronic eigenvalue. Together with `highest_occupied`, it defines the
+        electronic band gap.
+        """,
+    )
+
+    value_contributions = SubSection(
+        sub_section=BaseElectronicEigenvalues.m_def,
+        repeats=True,
+        description="""
+        Contributions to the electronic eigenvalues. Example, in the case of a DFT+GW calculation, the GW eigenvalues
+        are stored under `value`, and each contribution is identified by `label`:
+            - `'KS'`: Kohn-Sham contribution. This is also stored in the DFT entry under `ElectronicEigenvalues.value`.
+            - `'KSxc'`: Diagonal matrix elements of the expectation value of the Kohn-Sahm exchange-correlation potential.
+            - `'SigX'`: Diagonal matrix elements of the exchange self-energy. This is also stored in the GW entry under `ElectronicSelfEnergy.value`.
+            - `'SigC'`: Diagonal matrix elements of the correlation self-energy. This is also stored in the GW entry under `ElectronicSelfEnergy.value`.
+            - `'Zk'`: Quasiparticle renormalization factors contribution. This is also stored in the GW entry under `QuasiparticleWeights.value`.
+        """,
+    )
+
+    reciprocal_cell = Quantity(
+        type=KSpace.reciprocal_lattice_vectors,
+        description="""
+        Reference to the reciprocal lattice vectors stored under `KSpace`. This reference is useful when resolving the Brillouin zone
+        for the front-end visualization.
+        """,
+    )
+
+    def __init__(
+        self, m_def: Section = None, m_context: Context = None, **kwargs
+    ) -> None:
+        super().__init__(m_def, m_context, **kwargs)
+        self.name = self.m_def.name
+
+    def validate_occupation(self, logger: BoundLogger) -> bool:
+        """
+        Validate `occupation` by checking if they exist and have the same shape as `value`.
+
+        Args:
+            logger (BoundLogger): The logger to log messages.
+
+        Returns:
+            (bool): True if the shape of `occupation` is the same as `value`, False otherwise.
+        """
+        if self.occupation is None or len(self.occupation) == 0:
+            logger.warning('Cannot find `occupation` defined.')
+            return False
+        if self.value is not None and self.value.shape != self.occupation.shape:
+            logger.warning(
+                'The shape of `value` and `occupation` are different. They should have the same shape.'
+            )
+            return False
+        return True
```

Comment:
There are other things going on with `value` and its `__setattr__` defined in `PhysicalProperty`, so defining `shape` will not have any effect.

I just want to double check that `occupation` and `value` has the same shape, but it is true that this could maybe be a decorator, as it can become a quite popular check?

Reply:
I think a lot of our checks could be recast into decorators. I was already looking into it, but I'd have to gather all use cases. This is cleaning up for later.